libexec/rtld-elf/aarch64/reloc.c

```diff
@@ -167,6 +177,224 @@ _rtld_relocate_nonplt_self(Elf_Dyn *dynp, Elf_Auxinfo *aux)
+	asm ("msr	rctpidr_el0, %0" :: "C" (tls));
+
+	tls = xcalloc(DEFAULT_STACK_TABLE_SIZE, sizeof(*tls));
+	tls[0] = (uintptr_t)get_rstk;
+	asm ("msr	ctpidr_el0, %0" :: "C" (tls));
+
+	thr_thread_start(curthread);
+}
+
+void
+_rtld_thread_start_init(void (*p)(struct pthread *))
+{
+	assert(!(cheri_getperm(p) & CHERI_PERM_EXECUTIVE));
+	assert(!thr_thread_start);
+	thr_thread_start = tramp_pgs_append(p, obj_from_addr(p));
+}
+
+static void (*thr_sighandler)(int, siginfo_t *, void *);
+
+void
+_rtld_sighandler(int sig, siginfo_t *info, void *_ucp)
+{
+	uintptr_t csp, rcsp;
+	ucontext_t *ucp = _ucp;
+
+	csp = ucp->uc_mcontext.mc_capregs.cap_sp;
+	asm ("mrs	%0, rcsp_el0" : "=C" (rcsp));
+	ucp->uc_mcontext.mc_capregs.cap_sp = rcsp;
+
+	thr_sighandler(sig, info, ucp);
+
+	rcsp = ucp->uc_mcontext.mc_capregs.cap_sp;
+	asm ("msr	rcsp_el0, %0" :: "C" (rcsp));
+	ucp->uc_mcontext.mc_capregs.cap_sp = csp;
+}
+
+void
+_rtld_sighandler_init(void *p)
+{
+	assert(!(cheri_getperm(p) & CHERI_PERM_EXECUTIVE));
+	assert(!thr_sighandler);
+	thr_sighandler = tramp_pgs_append(p, obj_from_addr(p));
+}
+
+static void *
+set_bounds_get_remainder(struct tramp **inout, size_t len)
+{
+	void *hi = *inout;
+	void *lo;
+	ptraddr_t top_lo, alignment;
+	size_t remaining;
+
+	*inout = lo = cheri_setbounds(hi, len);
+	top_lo = cheri_gettop(lo);
+	remaining = cheri_gettop(hi) - top_lo;
+	alignment = CHERI_REPRESENTABLE_ALIGNMENT(remaining);
+	return (cheri_setaddress(hi, roundup2(top_lo, alignment)));
+}
+
+static int
+tramp_pg_create(struct tramp_pg **out)
+{
+	struct tramp_pg *p;
+	p = mmap(NULL,
+		 getpagesize(),
+		 PROT_READ | PROT_WRITE | PROT_EXEC,
+		 MAP_ANON | MAP_PRIVATE,
+		 -1, 0);
+	if (p == MAP_FAILED)
+		return (-1);
+	p->cursor = p->trampolines;
+	*out = p;
+	return (0);
+}
+
+void *
+tramp_pgs_append(const void *target, const Obj_Entry *dst)
+{
+	extern const struct tramp tramp_template_exe, tramp_template_res;
+
+	static struct tramp_pgs pgs = SLIST_HEAD_INITIALIZER(pgs);
+
+	const struct tramp *template;
+	struct tramp *t;
+	size_t len;
+
+	int n_retry = 0;
+	struct tramp_pg *pg = SLIST_FIRST(&pgs);
+
+	if (!pg) {
+retry:
+		if (n_retry++)
+			rtld_die();
+		if (tramp_pg_create(&pg))
+			rtld_die();
+		SLIST_INSERT_HEAD(&pgs, pg, entries);
+	}
+
+	if (cheri_getperm(target) & CHERI_PERM_EXECUTIVE)
+		template = &tramp_template_exe;
+	else
+		template = &tramp_template_res;
+	len = cheri_getlen(template);
+
+	t = roundup2(pg->cursor, _Alignof(typeof(*t)));
+	pg->cursor = set_bounds_get_remainder(&t, len);
+	if (!cheri_gettag(t))
+		goto retry;
+
+	memcpy(t, template, len);
+	/*
+	 * Arm64's i-cache and d-cache are not coherent. So we need to clean the d-cache
+	 * and invalidate the i-cache. See https://community.arm.com/arm-community-blogs/b/architectures-and-processors-blog/posts/caches-and-self-modifying-code
+	 * We derive the start and end addresses from pg so that they are aligned to the
+	 * cache-line size, since pg is page aligned.
+	 */
```

Comment:
This doesn't need a long explanation for why the call is needed, certainly not a link to a random Arm forum post. Just state that it's to ensure I/D-cache coherency, and give a quick note _why_ the rederivation for cache line alignment is needed.